// @serenity-js/protractor/lib/screenplay/models/ProtractorCookie.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProtractorCookie = void 0;
const core_1 = require("@serenity-js/core");
const web_1 = require("@serenity-js/web");
const tiny_types_1 = require("tiny-types");
const promised_1 = require("../promised");
/**
 * Protractor-specific implementation of {@apilink Cookie}.
 *
 * @group Models
 */
class ProtractorCookie extends web_1.Cookie {
    constructor(browser, cookieName) {
        super(cookieName);
        this.browser = browser;
        (0, tiny_types_1.ensure)('browser', browser, (0, tiny_types_1.isDefined)());
    }
    async delete() {
        return await (0, promised_1.promised)(this.browser.manage().deleteCookie(this.cookieName));
    }
    /**
     * @protected
     */
    async read() {
        const cookie = await this.browser.manage().getCookie(this.cookieName);
        if (!cookie) {
            throw new web_1.CookieMissingError(`Cookie '${this.cookieName}' not set`);
        }
        return {
            name: cookie.name,
            value: cookie.value,
            domain: cookie.domain,
            path: cookie.path,
            expiry: typeof cookie.expiry === 'number' && cookie.expiry >= 0
                ? core_1.Timestamp.fromTimestampInSeconds(Math.round(cookie.expiry))
                : undefined,
            httpOnly: cookie.httpOnly,
            secure: cookie.secure,
        };
    }
}
exports.ProtractorCookie = ProtractorCookie;
