// @serenity-js/protractor/lib/screenplay/models/locators/ProtractorRootLocator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProtractorRootLocator = void 0;
const web_1 = require("@serenity-js/web");
const promised_1 = require("../../promised");
/**
 * Protractor-specific implementation of {@apilink RootLocator}.
 *
 * @group Models
 */
class ProtractorRootLocator extends web_1.RootLocator {
    constructor(browser) {
        super();
        this.browser = browser;
    }
    async isPresent() {
        return true;
    }
    async nativeElement() {
        return {
            element: this.browser.element.bind(this.browser),
            all: this.browser.element.all.bind(this.browser),
        };
    }
    async switchToFrame(element) {
        // https://github.com/angular/protractor/issues/1846#issuecomment-82634739;
        const webElement = await element.getWebElement();
        await (0, promised_1.promised)(this.browser.switchTo().frame(webElement));
    }
    async switchToParentFrame() {
        await (0, promised_1.promised)(this.browser.driver.switchToParentFrame());
    }
    async switchToMainFrame() {
        await (0, promised_1.promised)(this.browser.driver.switchTo().defaultContent());
    }
}
exports.ProtractorRootLocator = ProtractorRootLocator;
