// @serenity-js/web/lib/stage/crew/photographer/Photographer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Photographer = void 0;
const core_1 = require("@serenity-js/core");
const events_1 = require("@serenity-js/core/lib/events");
const strategies = __importStar(require("./strategies"));
/**
 * The Photographer is a {@apilink StageCrewMember} who takes screenshots
 * using the web browser associated with the {@apilink Actor} that is currently {@apilink actorInTheSpotlight|in the spotlight}.
 *
 * ## Programmatically assigning the `Photographer` to the {@apilink Stage}
 *
 * ```ts
 * import { configure, ArtifactArchiver } from '@serenity-js/core'
 * import { Photographer, TakePhotosOfFailures } from '@serenity-js/web'
 *
 * configure({
 *   crew: [
 *     ArtifactArchiver.storingArtifactsAt(process.cwd(), 'target/site/serenity'),
 *     Photographer.whoWill(TakePhotosOfFailures),
 *   ]
 * })
 * ```
 *
 * ## Using `Photographer` with WebdriverIO
 *
 * ```ts
 * // wdio.conf.ts
 * import { ArtifactArchiver } from '@serenity-js/core'
 * import { Photographer, TakePhotosOfFailures } from '@serenity-js/web'
 * import { WebdriverIOConfig } from '@serenity-js/webdriverio'
 *
 * export const config: WebdriverIOConfig= {
 *
 *   // Tell WebdriverIO to use Serenity/JS framework
 *   framework: '@serenity-js/webdriverio',
 *
 *   serenity: {
 *     // Configure Serenity/JS to use an appropriate test runner adapter
 *     runner: 'cucumber',
 *     // runner: 'mocha',
 *     // runner: 'jasmine',
 *
 *     // register custom Actors class to configure your Serenity/JS actors
 *     actors: new Actors(),
 *
 *     // Register StageCrewMembers we've imported at the top of this file
 *     crew: [
 *       ArtifactArchiver.storingArtifactsAt(process.cwd(), 'target/site/serenity'),
 *       Photographer.whoWill(TakePhotosOfFailures),
 *     ]
 *   },
 *
 *   // ... rest of the config omitted for brevity
 * }
 * ```
 *
 * ## Using `Photographer` with Playwright Test
 *
 * ```ts
 * // playwright.config.ts
 * import type { PlaywrightTestConfig } from '@serenity-js/playwright-test'
 *
 * const config: PlaywrightTestConfig = {
 *     use: {
 *         crew: [
 *             [ '@serenity-js/web:Photographer', { strategy: 'TakePhotosOfFailures' } ] // or 'TakePhotosOfInteractions'
 *         ],
 *     },
 * };
 * export default config;
 * ```
 *
 * #### Learn more
 * - {@apilink SerenityOptions}
 *
 * ## Using `Photographer` with Protractor
 *
 * ```ts
 * // protractor.conf.js
 * const { ArtifactArchiver } = require('@serenity-js/core')
 * const { Photographer, TakePhotosOfFailures } = require('@serenity-js/web')
 *
 * exports.config = {
 *
 *   // Tell Protractor to use the Serenity/JS framework Protractor Adapter
 *   framework:      'custom',
 *   frameworkPath:  require.resolve('@serenity-js/protractor/adapter'),
 *
 *   serenity: {
 *         runner: 'jasmine',
 *         // runner: 'cucumber',
 *         // runner: 'mocha',
 *         crew: [
 *             ArtifactArchiver.storingArtifactsAt('./target/site/serenity'),
 *             Photographer.whoWill(TakePhotosOfFailures),
 *         ]
 *     },
 *
 *     // ... rest of the config omitted for brevity
 * }
 * ```
 *
 * ## Taking photos only upon failures only
 *
 * ```ts
 * import { Photographer, TakePhotosOfFailures } from '@serenity-js/web'
 *
 * Photographer.whoWill(TakePhotosOfFailures)
 * ```
 *
 * ## Taking photos of all the interactions
 *
 * ```ts
 * import { Photographer, TakePhotosOfInteractions } from '@serenity-js/web'
 *
 * Photographer.whoWill(TakePhotosOfInteractions)
 * ```
 *
 * ## Taking photos before and after all the interactions
 *
 * ```ts
 * import { Photographer, TakePhotosBeforeAndAfterInteractions } from '@serenity-js/web'
 *
 * Photographer.whoWill(TakePhotosBeforeAndAfterInteractions)
 * ```
 *
 * ## Learn more
 * - {@apilink Stage}
 * - {@apilink StageCrewMember}
 * - {@apilink TakePhotosBeforeAndAfterInteractions}
 * - {@apilink TakePhotosOfFailures}
 * - {@apilink TakePhotosOfInteractions}
 *
 * @group Stage
 */
class Photographer {
    /**
     * Instantiates a new {@apilink Photographer} configured to take photos (screenshots)
     * as per the specified {@apilink PhotoTakingStrategy}.
     *
     * @param strategy
     *  A no-arg constructor function that instantiates a {@apilink PhotoTakingStrategy}
     */
    static whoWill(strategy) {
        return new Photographer(new strategy());
    }
    /**
     * Instantiates a new {@apilink Photographer} configured to take photos (screenshots)
     * as per the specified {@apilink PhotoTakingStrategy}.
     *
     * @param config
     */
    static fromJSON(config) {
        if (config && config.strategy) {
            const availableStrategies = Object.keys(strategies).filter(strategy => strategy !== strategies.PhotoTakingStrategy.name); // not the abstract class
            if (availableStrategies.includes(config.strategy)) {
                return new Photographer(new strategies[config.strategy]());
            }
            throw new core_1.ConfigurationError(`'${config.strategy}' is not an available PhotoTakingStrategy. ` +
                `Available strategies: ${availableStrategies.join(', ')}.`);
        }
        return new Photographer(new strategies.TakePhotosOfFailures());
    }
    constructor(photoTakingStrategy, stage) {
        this.photoTakingStrategy = photoTakingStrategy;
        this.stage = stage;
    }
    /**
     * Assigns this {@apilink StageCrewMember} to a given {@apilink Stage}.
     *
     * @param stage
     *  An instance of a {@apilink Stage} this {@apilink StageCrewMember} will be assigned to
     */
    assignedTo(stage) {
        this.stage = stage;
        return this;
    }
    /**
     * Handles {@apilink DomainEvent} objects emitted by the {@apilink Stage}
     * this {@apilink StageCrewMember} is assigned to.
     *
     * @param event
     */
    notifyOf(event) {
        if (!this.stage) {
            throw new core_1.LogicError(`Photographer needs to be assigned to the Stage before it can be notified of any DomainEvents`);
        }
        if (!this.stage.theShowHasStarted()) {
            return void 0;
        }
        if (event instanceof events_1.ActivityStarts || event instanceof events_1.ActivityFinished) {
            this.photoTakingStrategy.considerTakingPhoto(event, this.stage);
        }
    }
}
exports.Photographer = Photographer;
