// tiny-types/lib/objects/deprecated.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.deprecated = void 0;
/**
 * @desc A decorator to mark a class, method or function as deprecated and make it log a warning whenever it is used.
 * Please see the tests for examples of usage.
 *
 * @param {string} message - describes the alternative implementation that should be used instead
 *                  of the deprecated method/function/class
 * @param {Logger} log - a function that handles the printing of the message,
 *                  such as {@link console.warn}
 */
function deprecated(message, log) {
    if (message === void 0) { message = ''; }
    if (log === void 0) { log = console.warn; }
    // eslint-disable-next-line unicorn/consistent-function-scoping,no-prototype-builtins
    var hasPrototype = function (target) { return target.hasOwnProperty('prototype'); };
    return function (target, propertyKey, descriptor) {
        if (target && propertyKey && descriptor) {
            return deprecateMethod(message, target, propertyKey, descriptor, log);
        }
        else if (hasPrototype(target)) {
            return deprecateClass(message, target, log);
        }
        else {
            throw new Error("Only a class, method or function can be marked as deprecated. ".concat(typeof target, " given."));
        }
    };
}
exports.deprecated = deprecated;
function deprecateClass(message, target, log) {
    return /** @class */ (function (_super) {
        __extends(class_1, _super);
        function class_1() {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            log("".concat(target.name, " has been deprecated. ").concat(message).trim());
            return _super.apply(this, args) || this;
        }
        return class_1;
    }(target));
}
function deprecateMethod(message, target, propertyKey, descriptor, log) {
    var originalMethod = descriptor.value;
    descriptor.value = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        log("".concat(target.constructor.name, "#").concat(propertyKey, " has been deprecated. ").concat(message).trim());
        return originalMethod.apply(this, args);
    };
    return descriptor;
}
