// @serenity-js/core/lib/screenplay/time/abilities/ScheduleWork.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScheduleWork = void 0;
const abilities_1 = require("../../abilities");
const models_1 = require("../models");
/**
 * An {@apilink Ability} that enables an {@apilink Actor} to schedule a callback function
 * to be executed with a delay, or until some condition is met.
 *
 * Used internally by the {@apilink Interaction|interaction} to {@apilink Wait}.
 *
 * @experimental
 *
 * @group Time
 */
class ScheduleWork extends abilities_1.Ability {
    constructor(clock, interactionTimeout) {
        super();
        this.scheduler = new models_1.Scheduler(clock, interactionTimeout);
    }
    initialise() {
        this.scheduler.start();
    }
    isInitialised() {
        return this.scheduler.isRunning();
    }
    /**
     * @param callback
     * @param limits
     */
    repeatUntil(callback, limits) {
        return this.scheduler.repeatUntil(callback, limits);
    }
    waitFor(delay) {
        return this.scheduler.waitFor(delay);
    }
    discard() {
        this.scheduler.stop();
    }
}
exports.ScheduleWork = ScheduleWork;
