// @serenity-js/web/lib/screenplay/questions/Value.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Value = void 0;
const core_1 = require("@serenity-js/core");
const models_1 = require("../models");
/**
 * Uses the {@apilink Actor|actor's} {@apilink Ability|ability} to {@apilink BrowseTheWeb} to retrieve
 * the `value` attribute of a given {@apilink PageElement}.
 *
 * ## Example widget
 * ```html
 * <input type="text" id="username" value="Alice" />
 * ```
 *
 * ## Retrieve the `value` of a given {@apilink PageElement}
 *
 * ```ts
 * import { actorCalled } from '@serenity-js/core'
 * import { Ensure, equals } from '@serenity-js/assertions'
 * import { By, PageElement, Value } from '@serenity-js/web'
 *
 * const usernameField = () =>
 *   PageElement.located(By.id('username'))
 *     .describedAs('username field')
 *
 * await actorCalled('Lisa')
 *   .attemptsTo(
 *     Ensure.that(Value.of(usernameField), equals('Alice')),
 *   )
 * ```
 *
 * ## Using Value as {@apilink QuestionAdapter}
 *
 * ```ts
 * import { actorCalled } from '@serenity-js/core'
 * import { Ensure, equals } from '@serenity-js/assertions'
 * import { By, PageElement, Value } from '@serenity-js/web'
 *
 * const usernameField = () =>
 *   PageElement.located(By.id('username'))
 *     .describedAs('username field')
 *
 * await actorCalled('Lisa')
 *   .attemptsTo(
 *     Ensure.that(
 *       Value.of(usernameField).toLocaleLowerCase()[0],
 *       equals('a')  // [a]lice
 *     ),
 *   )
 * ```
 *
 * ## Learn more
 * - {@apilink BrowseTheWeb}
 * - {@apilink MetaQuestion}
 * - {@apilink QuestionAdapter}
 * - {@apilink Question}
 *
 * @group Questions
 */
class Value extends core_1.Question {
    /**
     * Instantiates a {@apilink Question} that uses
     * the {@apilink Actor|actor's} {@apilink Ability|ability} to {@apilink BrowseTheWeb} to retrieve
     * the `value` attribute of a given {@apilink PageElement}.
     *
     * #### Learn more
     * - {@apilink MetaQuestion}
     *
     * @param pageElement
     */
    static of(pageElement) {
        return core_1.Question.createAdapter(new Value(pageElement));
    }
    constructor(element) {
        super();
        this.element = element;
        this.subject = (0, core_1.d) `the value of ${element}`;
    }
    /**
     * Instantiates a {@apilink Question} that uses
     * the {@apilink Actor|actor's} {@apilink Ability|ability} to {@apilink BrowseTheWeb} to retrieve
     * the `value` attribute of a given {@apilink PageElement}
     * located within the `parent` element.
     *
     * #### Learn more
     * - {@apilink MetaQuestion}
     *
     * @param parent
     */
    of(parent) {
        return new Value(models_1.PageElement.of(this.element, parent));
    }
    /**
     * @inheritDoc
     */
    async answeredBy(actor) {
        const element = await actor.answer(this.element);
        return element.value();
    }
    /**
     * @inheritDoc
     */
    describedAs(subject) {
        this.subject = subject;
        return this;
    }
    /**
     * @inheritDoc
     */
    toString() {
        return this.subject;
    }
}
exports.Value = Value;
